native/python/pyjp_buffer.cpp

```diff
@@ -0,0 +1,139 @@
+#include "jpype.h"
+#include "pyjp.h"
+#include "jp_buffer.h"
+#include "jp_buffertype.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+static void PyJPBuffer_dealloc(PyJPBuffer *self)
+{
+	JP_PY_TRY("PyJPBuffer_dealloc");
+	delete self->m_Buffer;
+	Py_TYPE(self)->tp_free(self);
+	JP_PY_CATCH();
+}
+
+static PyObject *PyJPBuffer_repr(PyJPBuffer *self)
+{
+	JP_PY_TRY("PyJPBuffer_repr");
+	JPContext *context = PyJPModule_getContext();
+	JPJavaFrame frame(context);
+	if (self->m_Buffer == NULL)
+		JP_RAISE(PyExc_ValueError, "Null array");
+	stringstream sout;
+
+	sout << "<java beffer " << self->m_Buffer->getClass()->toString() << ">";
+	return JPPyString::fromStringUTF8(sout.str()).keep();
+	JP_PY_CATCH(0);
+}
+
+static void PyJPBuffer_releaseBuffer(PyJPBuffer *self, Py_buffer *view)
+{
+	JP_PY_TRY("PyJPBufferPrimitive_releaseBuffer");
+	JP_PY_CATCH();
+}
+
+int PyJPBuffer_getBuffer(PyJPBuffer *self, Py_buffer *view, int flags)
+{
+	JP_PY_TRY("PyJPBufferPrimitive_getBuffer");
+	JPContext *context = PyJPModule_getContext();
+	JPJavaFrame frame(context);
+	if (self->m_Buffer == NULL)
+		JP_RAISE(PyExc_ValueError, "Null buffer");
+	try
+	{
+		JPBuffer *buffer = self->m_Buffer;
+
+		if (!buffer->isValid())
+		{
+			PyErr_SetString(PyExc_BufferError, "Java buffer is not direct");
+			return -1;
+		}
+
+		if ((flags & PyBUF_WRITEABLE) == PyBUF_WRITEABLE && buffer->isReadOnly())
+		{
+			PyErr_SetString(PyExc_BufferError, "Java buffer is not writable");
+			return -1;
+		}
+
+		*view = buffer->getView();
+
+		// If strides are not requested and this is a slice then fail
+		if ((flags & PyBUF_STRIDES) != PyBUF_STRIDES)
+		{
+			if (view->strides[0] != view->itemsize)
+				JP_RAISE(PyExc_BufferError, "slices required strides");
+			view->strides = NULL;
+		}
+
+		// If shape is not requested
+		if ((flags & PyBUF_ND) != PyBUF_ND)
+		{
+			view->shape = NULL;
+		}
+
+		// If format is not requested
+		if ((flags & PyBUF_FORMAT) != PyBUF_FORMAT)
+			view->format = NULL;
+
+		// Okay all successful so reference the parent object
+		view->obj = (PyObject*) self;
+		Py_INCREF(view->obj);
+		return 0;
+	} catch (JPypeException &ex)
+	{
+		PyJPBuffer_releaseBuffer(self, view);
```

Comment:
why call the stub here, if it does nothing?

Reply:
This was left from the array pattern where it is required.  I kept the pattern in case there was actually some need to call the stub.  I can remove it if you think that the stub will never be required.

Reply:
If it does no harm, lets just leave it. Eventually it will be needed for whatever reason.